docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,739 @@
+{
```

Comment:
<p>I think we can/should update the best discord candidate on the fly:</p><p> </p><pre class="ql-syntax" spellcheck="false">def _find_discord(subseqs, cands):
&nbsp; &nbsp; n, m = subseqs.shape
&nbsp; &nbsp; excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))

    discord_idx = -1
    discord_R = np.inf
    discord_nn_idx = -1

    for idx in cands:
        R = np.matmul(subseqs, subseqs[i])
&nbsp; &nbsp; &nbsp; &nbsp; R[max(0, i - excl_zone) : min(n, i + excl_zone + 1)] = np.NINF
        nn_idx = np.argmax(R)
        if R[idx] &lt; discord_R:
            discord_R = R[idx]
            discord_idx = i
&nbsp; &nbsp; &nbsp; &nbsp;     discord_nn_idx = idx

&nbsp; &nbsp; discord_dist = np.sqrt(2 * m * (1 - (discord_R / m)))
&nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; return discord_idx, discord_dist, discord_nn_idx
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Reply:
I tried it. However, then I realized  we cannot easily return `top-k` discords if we decide to do so in the future using this approach. 

For instance, say, we want to return `top-2` discords. After getting the `top-1` discord (located at `top-1-idx`), we should remove all the candidates located in its `exclusion area`. And then, search for the second discord among the rest of candidates. If there is None, we have to go back and lower the threshold and do the first two phases again.

What do you think?

Reply:
<p>Do you know if top-k is only assessed in <code>_find_discord</code>  ? Or will it affect the first two phases as well?</p><p> </p><p>So, the traditional way of doing this is to use <code>heapq</code>  but, lately, I have been thinking about another idea using <code>np.searchsorted</code>  that should be pretty straightforward to implement:</p><p> </p><p>https://stackoverflow.com/questions/68037795/find-top-k-smallest-values-so-far-in-data-stream</p>

Reply:
<p>Also, I'm not sure what is going on with the code formatting. I keep on fixing it but it still looks indented in the wrong places</p>

Reply:
> Do you know if top-k is only assessed in `_find_discord` ? Or will it affect the first two phases as well?

If we realize the chosen candidates (at the end of phase two, i.e. output of `_prune_candidates`) is not **enough (see below)** for k discords, we have to lower the `min_dist` (so more will be selected in the first phase and less will be pruned in the second phase) and do the first two phases again. In other words, we are not entering the third phase unless we are sure we have got enough candidates 

**An extreme scenario like below might help us to better understand this:** 
What if I want to return k discords where `k=len(T) - m + 1` (i.e. all subsequences!!!)? In that case, if I start at a large `min_dist` and realize the number of candidates at the end of second phase is not enough, I need to go through the first two phases again but will lower 'min_dist'. If I set `min_dist` to a VERY SMALL value,  I will probably get all subsequences as the candidates. Perfect! 

---

What is **enough** number of candidates? 

The authors of MERLIN paper cited [this paper](https://www.cs.ucr.edu/~eamonn/DiskawareDiscords.pdf) when they discussed `top-k` discords. I took a look at the paper and it says:

> Furthermore, it may also be the case that the users would like to detect the top k discords regardless
of the distance to their neighbors. In those cases, specifying a large threshold will result in an empty set, while a very
small range parameter may have high time and space complexity. With this in mind, a reasonable strategy to detect the top k discords would be to start with a “relatively large” threshold and if in the end |C| < k, to restart the algorithm with a smaller parameter.
(C: set of candidates)

Therefore, according to this paper, we just need to get at least `k` candidates (in the first two phases) before jumping into the third phase. 

---

**Another opinion on "enough"**
I think I saw a different approach about "how" to choose `top-k` discords before. Unfortunately I cannot find it in the MERLIN support website anymore. I went through some of `.m` files but couldn't find it. Maybe I am wrong (and this is just an illusion!?) or maybe it is just removed from the website. I briefly explain that opinion below:

In the third phase `_find_discord`: If we find the top discord located at index `idx`, we need to first remove all the candidates that are in its exclusion area. In other words, any candidate with starting index `[idx-excl_zone : idx + excl_zone+1]` should be removed (as it might not make sense to discover top-2 discords that are just located next to each other!). Then, we need to find the next discord from the  remaining ones. In this case, I think we need to go through the phase 3 and in the middle of the way, we might have to break it and go through the first two phases!!! I will try to see if I can find this in their MATLAB code. 

---

**Final thought:** we can just go with what the paper stated and just check the number of candidates at the end of phase 2. If we realize it is less than `k`, we can start over and do the first two phases again with lower `min_dist`. I think it is based on a peer-reviewed paper and its implementation is easier and its running time will be lower.

Reply:
@seanlaw  
Update:
I took a closer look at the MATLAB file `MERLIN_topK`. I got it. I will explain it in the notebook as I don't want to make this PR messy. Is that okay?